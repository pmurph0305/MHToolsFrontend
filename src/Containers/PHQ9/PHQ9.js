import React from "react";
import Table from "./Table/Table";
import SectionInfo from "../../Components/SectionInfo/SectionInfo";
import SelectionBox from "../../Components/SelectionBox/SelectionBox";

import "./PHQ9.scss";

// Headers for the PHQ-9.
const phq9Headers = [
  "Over the past 2 weeks, how often have you been bothered by any of the following problems?",
  "Select an answer"
];

// Standard PHQ-9 questions.
const phq9Questions = [
  "1. Little interest or pleasure in doing things",
  "2. Feeling down, depressed, or hopeless",
  "3. Trouble falling asleep, staying asleep, or sleeping too much",
  "4. Feeling tired or having little energy",
  "5. Poor appetite or overeating",
  "6. Feeling bad about yourself - or that you're a failure or have let yourself or your family down",
  "7. Trouble concentrating on things, such as reading the newspaper or watching television",
  "8. Moving or speaking so slowly that other people could have noticed. Or, the opposite - being so fidgety or restless that you have been moving around a lot more than usual",
  "9. Thoughts that you would be better off dead or of hurting yourself in some way",
  "10. If you checked off any problems, how difficult have those problems made it for you to do your work, take care of things at home, or get along with other people?"
];

// Standard selection options for PHQ-9 Questions 1-9.
const phq9SelectionOptions = [
  "Not at all",
  "Several days",
  "More than half days",
  "Nearly every day"
];

// Selection options for PHQ-9 Question 10.
const phq9Q10SelectionOptions = [
  "Not difficult at all",
  "Somewhat difficult",
  "Very difficult",
  "Extremely difficult"
];

// Initial state of the PHQ-9 Answers.
const initialState = {
  answers: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
};

class PHQ9 extends React.Component {
  constructor(props) {
    super(props);
    this.state = initialState;
  }

  // Handles submitting the phq9 when the submit button is clicked.
  onSubmitPHQ9 = () => {
    this.props.onSubmitPHQ9(this.state.answers);
  };

  // Handles updating state when a question is answered.
  onSelectAnswer = event => {
    // get current state's answers
    let answers = this.state.answers;
    // update proper index (event.target.id is the correct index)
    answers[event.target.id] = Number(event.target.value);
    // use object.assign to assign the new array to update the state's answers.
    this.setState(Object.assign(this.state.answers, answers));
  };

  // Builds the footers based on the current state of submission.
  getFooters = () => {
    return [
      !this.props.submissionResult ? (
        ""
      ) : this.props.submissionResult === "PHQ-9 submission successful." ? (
        <p className="SubmissionSuccess">{this.props.submissionResult}</p>
      ) : (
        <p className="SubmissionFailed">{this.props.submissionResult}</p>
      ),
      <button className="SubmitButton" onClick={this.onSubmitPHQ9}>
        Submit
      </button>
    ];
  };

  // Builds an array of the html elements for the PHQ-9 table.
  getRows = () => {
    return phq9Questions.map((question, index) => {
      let selection;
      if (index < 9) {
        selection = (
          <SelectionBox
            id={index}
            key={index}
            label={"select answer for question " + (index + 1)}
            value={this.state.answers[index]}
            options={phq9SelectionOptions}
            onChange={this.onSelectAnswer}
          />
        );
      } else {
        selection = (
          <SelectionBox
            id={index}
            key={index}
            label={"select answer for question " + (index + 1)}
            value={this.state.answers[index]}
            options={phq9Q10SelectionOptions}
            onChange={this.onSelectAnswer}
          />
        );
      }
      return [question, selection];
    });
  };

  render() {
    return (
      <section className="PHQ9Section">
        <SectionInfo
          title={"PHQ-9"}
          description={
            "The Patient Health Questionnaire (PHQ-9) is a multipurpose tool used for screening, diagnosing, and monitoring the severity of depression."
          }
        />
        <Table
          tClass="TableClass"
          trClass="TableRow"
          tdClass="TableCell"
          thClass="TableHeader"
          tfClass="TableFooter"
          headers={phq9Headers}
          rows={this.getRows()}
          footers={this.getFooters()}
        />
        <p>{this.state.total}</p>
      </section>
    );
  }
}

export default PHQ9;
