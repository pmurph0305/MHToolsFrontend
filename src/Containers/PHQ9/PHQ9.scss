@import '../../styles/shared.scss';

.PHQ9Section {
    margin: 0;
    color: rgba(0, 0, 0, 0.90);
    text-align: center;
    background-color: $primary-background;
    padding: 0.5rem;
}

.PHQ9Title {
    margin: 0.25rem 0.5rem;
}

.PHQ9Description {
    padding: 1rem;
}

.TableClass {
    font-weight: 600;
    margin: auto;
    line-height: 1.5;
    text-align: left;
    width: 100%;
    max-width: 75rem;
    padding: 1rem;

    // Striping every other row for easier visibilty.
    .TableRow:nth-child(even) {
        background-color: $secondary-background;
    }

    %TableCell-shared {
        padding: 1rem;
        border-bottom: 1px solid black;
        border-right: 1px solid black;
        &:first-child{ border-left: 1px solid black; }
    }

    .TableCell {
        @extend %TableCell-shared;
    }

    .TableFooter {
        @extend %TableCell-shared;
        text-align: right;
        // Align button center.
        &:last-child{ text-align: center }
    }
    
    .TableHeader {
        @extend %TableCell-shared;
        background-color: $secondary-background;
        border-top: 1px solid black;
        font-weight: bold;
    }
}

%SubmissionResult-shared {
    text-align: right;
    display: inline-block;
    font-weight: 600;
    margin: 0;
    padding: 0.25rem;
    color: $primary-text-color;
}

.SubmissionSuccess {
    @extend %SubmissionResult-shared;
    background-color: #73c7a8;
}

.SubmissionFailed {
    @extend %SubmissionResult-shared;
    background-color:  #ff9999;
}

.SubmitButton {
    @extend %ButtonBase;

    // Large screens
    @media screen and (min-width:60em) {
        padding-left: 4rem;
        padding-right: 4rem;
    }
    // Medium
    @media screen and (min-width:30em) and (max-width: 60em) {
        padding-left: 2rem;
        padding-right: 2rem;
    }
    // Small
    @media screen and (max-width: 30em) {
        padding-left: 0.5rem;
        padding-right:0.5rem;
    }

}