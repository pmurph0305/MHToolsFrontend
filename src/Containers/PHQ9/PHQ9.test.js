// test file
import { shallow, mount } from 'enzyme';
import React from 'react';
import PHQ9 from './PHQ9'


describe('PHQ9 tests', () => {

    it('renders PHQ9 snapshot correctly', () => {
        const wrapper = shallow(<PHQ9 />);
        expect(wrapper).toMatchSnapshot();
    })

    it('submits phq9 correctly', () => {
        const mockSubmit = jest.fn();
        const wrapper = mount(<PHQ9 onSubmitPHQ9={mockSubmit}/>)
        wrapper.find('button').simulate('click');
        expect(mockSubmit).toHaveBeenCalledTimes(1);
        wrapper.unmount();
    })

    it('gets footers correctly', () => {
        expect.assertions(6);
        const wrapFooters = shallow(<PHQ9 />)
        let footers = wrapFooters.instance().getFooters();
        expect(footers[0]).toEqual('');
        expect(footers[1].type).toEqual('button');

        const wrapFooters2 = shallow(<PHQ9 submissionResult='PHQ-9 submission successful.'/>);
        let footers2 = wrapFooters2.instance().getFooters();
        expect(footers2[0].props.children).toEqual('PHQ-9 submission successful.');
        expect(footers2[1].type).toEqual('button');

        const wrapFooters3 = shallow(<PHQ9 submissionResult='PHQ-9 submission test fail'/>);
        let footers3 = wrapFooters3.instance().getFooters();
        expect(footers3[0].props.children).toEqual('PHQ-9 submission test fail');
        expect(footers3[1].type).toEqual('button');
    })

    it('selects and updates answers correctly', () => {
        const wrapper2 = mount(<PHQ9/>);
        expect(wrapper2.exists()).toBe(true);
        const initialAnswers = [0,0,0,0,0,0,0,0,0,0];
        expect(wrapper2.state('answers')).toEqual(initialAnswers);
        expect(wrapper2.find('select')).toHaveLength(10);
        const mockEvent = {
            target: {
                id: 0,
                value: 2
            }
        }
        const newAnswers = [2,0,0,0,0,0,0,0,0,0];
        wrapper2.find('select').first().simulate('change', mockEvent);
        expect(wrapper2.state('answers')).toEqual(newAnswers)
    })
})