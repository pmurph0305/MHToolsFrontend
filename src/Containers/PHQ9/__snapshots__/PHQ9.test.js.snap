// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PHQ9 tests renders PHQ9 snapshot correctly 1`] = `ShallowWrapper {}`;
